README
:: Energy ::

Energy system for social games.

The documentation is available at
http://packages.python.org/energy

by Heungsub Lee


CHANGES
Version 0.1.9
-------------

Released on June 24th 2014.

Fixes a bug on restored extra energy. This issue was reported by
`@hest <https://github.com/hest>`_.

Version 0.1.8
-------------

Released on May 13th 2014.

- Adds ``force`` option for :meth:`Energy.use` to use energy even if there is
  not enough energy.
- Adds :meth:`Energy.debt`.

(by `@youknowone <https://github.com/youknowone>`_)

Version 0.1.7
-------------

Released on Nov 5th 2013.

``recovery_quantity`` works. Thanks to `Donggeun Lee <https://github.com/
incl>`_. (commit `c98f348 <https://github.com/sublee/energy/commit/c98f348
bed85522e491a6427e45e6ed527396eea>`_)

Version 0.1.6
-------------

Released on Jan 21st 2013.

``used_at`` parameter is allowed to be a ``datetime`` object or timestamp
number.

Version 0.1.5
-------------

Released on Jan 11th 2013.

- Adds :meth:`Energy.recover_fully_in` to calculate seconds to be recovered
  fully. It is useful for making push notification.
- Supports Python 3 and Jython.

Version 0.1.4
-------------

Released on Dec 17th 2012.

Fixes a bug about the timezone of the default ``time`` argument. The
documentation describes the timezone as UTC but it was a local timezone.

If your server is already using this project and setted with non-UTC timezone,
you should override :func:`timestamp` to use local timezone.

Version 0.1.3
-------------

Released on Dec 11th 2012.

- Adds ``future_tolerance`` parameter to be tolerant of using energy at the
  future. It is possible if there are several servers.
- Changes a form of pickled :class:`Energy` from ``tuple`` to ``dict``.
- Supports Jython.

Version 0.1.2
-------------

Released on Oct 12th 2012.

- Fixes an error when the energy has :attr:`Energy.used_at` before
  initialization.
- Fixes an error when using specific time for :meth:`Energy.use`.
- :class:`Energy` can be compared with number (`<`, `<=`, `>`, `>=`).

Version 0.1.1
-------------

Released on Sep 25th 2012.

- Supports the argumented arithmetic assignment (`+=`, `-=`).
- Fixes a calculation error when consuming energy over the maximum.

Version 0.1
-----------

First public preview release.


energytests.py
# -*- coding: utf-8 -*-
from __future__ import with_statement
from calendar import timegm
from contextlib import contextmanager
from datetime import datetime, timedelta
from functools import partial
from time import gmtime

from pytest import raises

from energy import Energy, timestamp


@contextmanager
def time_traveler():
    import energy
    changed_time = [0]
    def T(time=None):
        if time is None:
            return changed_time[0]
        else:
            changed_time[0] = time
    original_timestamp = energy.timestamp
    energy.timestamp = partial(energy.timestamp, default_time_getter=T)
    yield T
    energy.timestamp = original_timestamp


def test_timestamp():
    _1sec = timedelta(0, 1)
    assert datetime.utcfromtimestamp(timestamp()) - datetime.utcnow() < _1sec
    assert datetime.fromtimestamp(timestamp()) - datetime.now() < _1sec
    assert timestamp() - timegm(gmtime()) < 1


def test_init_energy():
    energy = Energy(10, 10)
    assert isinstance(energy.recovery_interval, int)
    assert energy.recovery_interval == 10
    energy = Energy(10, timedelta(seconds=10))
    assert isinstance(energy.recovery_interval, (int, float))
    assert energy.recovery_interval == 10


def test_use_energy():
    energy = Energy(10, 1000)
    assert energy == 10
    energy.use()
    assert energy == 9
    energy.use(5)
    assert energy == 4
    with raises(ValueError):
        energy.use(5)
    energy.use(10, timestamp() + 10000)
    assert energy.current(timestamp() + 10000) == 0
    with raises(ValueError):
        energy.use(10, timestamp() + 10010)
    energy.use(10, timestamp() + 10010, force=True)
    assert energy.recover_in(timestamp() + 10010) == 11000
    assert energy.current(timestamp() + 10010) == 0
    assert energy.debt(timestamp() + 10010) == 10
    energy.use(10, timestamp() + 10010, force=True)
    assert energy.recover_in(timestamp() + 10010) == 21000
    assert energy.debt(timestamp() + 10010) == 20


def test_set_energy():
    energy = Energy(10, 1000)
    energy.set(1)
    assert energy == 1
    energy.set(5)
    assert energy == 5


def test_reset_energy():
    energy = Energy(10, 1000)
    energy.use(5)
    assert energy == 5
    energy.reset()
    assert energy == 10


def test_cast_energy():
    true_energy = Energy(1, 1000)
    false_energy = Energy(0, 1000)
    assert int(true_energy) == 1
    assert int(false_energy) == 0
    assert float(true_energy) == 1.0
    assert float(false_energy) == 0.0
    assert bool(true_energy) is True
    assert bool(false_energy) is False


def test_recover_energy():
    energy = Energy(10, 5)
    with time_traveler() as T:
        T(0)
        energy.use(2)
        T(1)
        assert energy == 8
        assert energy.recover_in() == 4
        assert energy.recover_fully_in() == 9
        T(2)
        assert energy == 8
        assert energy.recover_in() == 3
        assert energy.recover_fully_in() == 8
        T(3)
        assert energy == 8
        assert energy.recover_in() == 2
        assert energy.recover_fully_in() == 7
        T(4)
        assert energy == 8
        assert energy.recover_in() == 1
        assert energy.recover_fully_in() == 6
        T(5)
        assert energy == 9
        assert energy.recover_in() == 5
        assert energy.recover_fully_in() == 5
        T(9)
        assert energy == 9
        assert energy.recover_in() == 1
        assert energy.recover_fully_in() == 1
        T(10)
        assert energy == 10
        assert energy.recover_in() == None
        assert energy.recover_fully_in() == None
        T(100)
        assert energy == 10
        assert energy.recover_in() == None
        assert energy.recover_fully_in() == None


def test_use_energy_while_recovering():
    energy = Energy(10, 5)
    with time_traveler() as T:
        T( 0); energy.use(5)
        T( 1); assert energy == 5
        T( 2); energy.use(1)
        T( 3); assert energy == 4
        T( 4); assert energy == 4
        T( 5); assert energy == 5
        T( 6); assert energy == 5
        T( 7); energy.use(1)
        T( 8); assert energy == 4
        T( 9); assert energy == 4
        T(10); assert energy == 5


def test_use_energy_after_recovered():
    energy = Energy(10, 5)
    with time_traveler() as T:
        T( 0); energy.use(10)
        T( 1); assert energy == 0
        T( 5); energy.use(1)
        T( 6); assert energy == 0


def test_use_energy_at_the_future():
    energy = Energy(10, 5)
    with time_traveler() as T:
        T( 5); energy.use()
        T( 6); assert energy.passed() == 1
        with raises(ValueError):
            T( 4); energy.passed()
        with raises(ValueError):
            T( 3); energy.passed()
        with raises(ValueError):
            T( 2); energy.passed()
        with raises(ValueError):
            T( 1); energy.passed()
        with raises(ValueError):
            T( 0); energy.passed()


def test_future_tulerance():
    energy = Energy(10, 5, future_tolerance=4)
    with time_traveler() as T:
        T(5)
        energy.use()
        # used at the past
        T(6)
        assert energy.passed() == 1
        assert energy == 9
        # used at the near future
        T(4)
        assert energy.passed() == 0
        assert energy == 9
        T(3)
        assert energy.passed() == 0
        assert energy == 9
        T(2)
        assert energy.passed() == 0
        assert energy == 9
        T(1)
        assert energy.passed() == 0
        assert energy == 9
        # used at the remote future
        T(0)
        with raises(ValueError):
            energy.passed()


def test_pickle_energy():
    try:
        import cPickle as pickle
    except ImportError:
        import pickle
    energy = Energy(10, 5)
    with time_traveler() as T:
        T( 0); assert energy == 10
        T( 1); energy.use(5)
        T( 2); assert energy == 5
        dump = pickle.dumps(energy)
        loaded_energy = pickle.loads(dump)
        assert energy == loaded_energy
        T( 3); assert energy == 5
        T( 3); assert loaded_energy == 5


class OldEnergy(Energy):

    def __setstate__(self):
        return (self.max, self.recovery_interval, self.recovery_quantity,
                self.used, self.used_at)


def test_pickle_energy_compatibility():
    try:
        import cPickle as pickle
    except ImportError:
        import pickle
    energy = OldEnergy(10, 5)
    with time_traveler() as T:
        T( 0); assert energy == 10
        T( 1); energy.use(5)
        T( 2); assert energy == 5
        dump = pickle.dumps(energy)
        dump = dump.replace('energytests\nOldEnergy'.encode(),
                            'energy\nEnergy'.encode())
        loaded_energy = pickle.loads(dump)
        assert type(loaded_energy) is Energy
        T( 3); assert energy == 5
        T( 3); assert loaded_energy == 5


def test_save_and_retrieve_energy():
    energy = Energy(10, 5)
    with time_traveler() as T:
        T( 0); assert energy == 10
        T( 1); energy.use(5)
        T( 2); assert energy == 5
        T(3)
        saved = energy.used
        saved_used, saved_used_at = energy.used, energy.used_at
        T(11)
        assert energy == 7
        loaded_energy = Energy(10, 5, used=saved_used, used_at=saved_used_at)
        assert loaded_energy == 7
        assert loaded_energy == energy
        loaded_energy2 = Energy(10, 5)
        loaded_energy2.used = saved_used
        loaded_energy2.used_at = saved_used_at
        assert loaded_energy2 == 7
        assert loaded_energy2 == energy
        loaded_energy3 = object.__new__(Energy)
        loaded_energy3.__init__(10, 5, used=saved_used, used_at=saved_used_at)
        assert loaded_energy3 == 7
        assert loaded_energy3 == energy
        loaded_energy4 = object.__new__(Energy)
        loaded_energy4.used = saved_used
        loaded_energy4.used_at = saved_used_at
        loaded_energy4.__init__(10, 5, used=saved_used, used_at=saved_used_at)
        assert loaded_energy4 == 7
        assert loaded_energy4 == energy


def test_float_recovery_interval():
    energy = Energy(10, 0.5)
    with time_traveler() as T:
        T( 0); energy == 10
        T( 1); energy.use(3)
        T( 2); energy == 9
        T( 3); energy == 10


def test_equivalent_energy():
    assert Energy(10, 10) == Energy(10, 10)
    assert Energy(5, 10) != Energy(10, 10)
    e1, e2, e3 = Energy(10, 10), Energy(10, 10), Energy(8, 10)
    with time_traveler() as T:
        T(123)
        e1.use()
        e2.use()
        assert e1 == e2
        T(128)
        e1.use()
        assert e1 != e2
        assert int(e1) == int(e3)
        assert e1 != e3


def test_set_max_energy():
    energy = Energy(10, 300)
    with time_traveler() as T:
        T( 0); assert energy == 10
        T( 1); energy.max = 11
        T( 2); assert energy == 11
        T( 3); energy.use()
        T( 4); assert energy == 10
        T( 5); energy.max = 12
        T( 6); assert energy == 10
        T( 7); energy.max = 9
        T( 8); assert energy == 9
        T( 9); energy.max = 1
        T(10); assert energy == 1
        T(11); energy.max = 10
        T(12); assert energy == 10


def test_extra_energy():
    energy = Energy(10, 300)
    with time_traveler() as T:
        T(0)
        energy.set(15)
        T(1)
        assert energy == 15
        assert energy.recover_in() is None
        assert energy.recover_fully_in() is None
        T(2)
        energy.use()
        assert energy.recover_in() is None
        assert energy.recover_fully_in() is None
        T(6)
        energy.use(6)
        T(7)
        assert energy.recover_in() == 299
        assert energy.recover_fully_in() == 599
        T(8)
        assert energy.recover_in() == 298
        assert energy.recover_fully_in() == 598
        T(9)
        energy.set(15)
        assert energy.recover_in() is None
        assert energy.recover_fully_in() is None
        T(10)
        assert energy.recover_in() is None
        assert energy.recover_fully_in() is None


def test_restored_extra_energy():
    energy = Energy(10, 300, used=-5, used_at=0)
    assert energy.used_at is None
    with time_traveler() as T:
        T(0)
        assert energy == 15
        energy.use()
        assert energy == 14
    assert energy.used_at is None


def test_repr_energy():
    energy = Energy(10, 300)
    with time_traveler() as T:
        T( 0); assert repr(energy) == '<Energy 10/10>'
        T( 1); energy.use()
        T( 2); assert repr(energy) == '<Energy 9/10 recover in 04:59>'


def test_compare_energy():
    energy = Energy(10, 300)
    with time_traveler() as T:
        T(0)
        assert energy == 10
        assert energy > 9
        assert 9 < energy
        assert energy < 11
        assert 11 > energy
        assert 9 < energy < 11
        assert energy <= 10
        assert energy >= 10
        assert 10 <= energy
        assert 10 >= energy
        assert 10 <= energy <= 10


def test_arithmetic_assign_energy():
    energy = Energy(10, 3)
    with time_traveler() as T:
        T( 0); energy += 10
        T( 1); assert energy == 20
        T( 2); energy -= 13
        T( 3); assert energy == 7
        T( 6); assert energy == 8
        T( 7); energy += 10
        T( 8); energy -= 10;
        T( 9); assert energy.recover_in() == 2
        T(10); assert energy.recover_in() == 1
        T(11); assert energy == 9


def test_various_used_at():
    with time_traveler() as T:
        T(2)
        energy = Energy(10, 3, used=1, used_at=1)
        assert energy == 9
        T(5)
        assert energy == 10
        T(2)
        energy = Energy(10, 3, used=1, used_at=datetime.utcfromtimestamp(1))
        assert energy == 9
        T(5)
        assert energy == 10


def test_recovery_quantity():
    with time_traveler() as T:
        T(0)
        energy = Energy(10, 3, 2)
        assert energy == 10
        energy.use()
        energy.use()
        energy.use()
        assert energy == 7
        T(1)
        assert energy == 7
        T(2)
        assert energy == 7
        T(3)
        assert energy == 9
        T(6)
        assert energy == 10


energy.py
# -*- coding: utf-8 -*-
"""
    energy
    ~~~~~~

    Energy system for social games such as FarmVille or The Sims Social.

    :copyright: (c) 2012-2013 by Heungsub Lee
    :license: BSD, see LICENSE for more details.
"""
from calendar import timegm
from datetime import datetime, timedelta
import sys
from time import gmtime, struct_time


__version__ = '0.1.9'
__all__ = ['Energy']


def timestamp(time=None, default_time_getter=gmtime):
    """Makes some timestamp.

    1. If you pass a :class:`datetime` object, it makes a timestamp from the
       argument.
    2. If you pass a timestamp(`int` or `float`), it just returns that.
    3. If you call it without parameter, it makes a timestamp from the result
       of `default_time_getter`.
    """
    if time is None:
        time = default_time_getter()
    if isinstance(time, datetime):
        return timegm(time.timetuple())
    elif isinstance(time, struct_time):
        return timegm(time)
    return int(time)


if sys.version_info < (2, 6):
    # A fallback of property under Python 2.6. The code is from
    # http://blog.devork.be/2008/04/xsetter-syntax-in-python-25.html
    class property(property):
        def __init__(self, fget, *args, **kwargs):
            self.__doc__ = fget.__doc__
            super(property, self).__init__(fget, *args, **kwargs)
        def setter(self, fset):
            ns = sys._getframe(1).f_locals
            for k, v in ns.iteritems():
                if v == self:
                    propname = k
                    break
            ns[propname] = property(self.fget, fset, self.fdel, self.__doc__)
            return ns[propname]


if not hasattr(timedelta, 'total_seconds'):
    # A fallback of timedelta.total_seconds under Python 2.7 and Python 3.1.
    def total_seconds(timedelta):
        ms, s, d = timedelta.microseconds, timedelta.seconds, timedelta.days
        return (ms + (s + d * 24 * 3600) * (10 ** 6)) / (10 ** 6)


class Energy(object):
    """A consumable and recoverable stuff in social gamers. Think over
    reasonable energy parameters for your own game. Energy may decide return
    period of your players.

    :param max: maximum energy
    :param recovery_interval: an interval in seconds to recover energy
    :type recovery_interval: number or ``timedelta``
    :param recovery_quantity: a quantity of once energy recovery. Defaults to
                              ``1``.
    :param future_tolerance: near seconds to ignore exception when used at the
                             future
    :param used: set this when retrieve an energy, otherwise don't touch
    :param used_at: set this when retrieve an energy, otherwise don't touch
    :type used_at: timestamp number or ``datetime``

    :raise TypeError: some argument isn't valid type
    """

    #: Quantity of used energy.
    used = 0

    #: A time when using the energy first.
    used_at = None

    def __init__(self, max, recovery_interval, recovery_quantity=1,
                 future_tolerance=None, used=used, used_at=used_at):
        if not isinstance(max, int):
            raise TypeError('max should be int')
        if not isinstance(recovery_quantity, int):
            raise TypeError('recovery_quantity should be int')
        if isinstance(recovery_interval, timedelta):
            try:
                recovery_interval = recovery_interval.total_seconds()
            except AttributeError:
                recovery_interval = total_seconds(recovery_interval)
        if not isinstance(recovery_interval, (int, float)):
            raise TypeError('recovery_interval should be number')
        self._max = max
        #: The interval in seconds to recover energy.
        self.recovery_interval = recovery_interval
        #: The quantity of once energy recovery.
        self.recovery_quantity = recovery_quantity
        #: The near seconds to ignore exception when used at the future.
        #:
        #: .. versionadded:: 0.1.3
        self.future_tolerance = future_tolerance
        self.used = used
        if 0 < used and used_at is not None:
            self.used_at = timestamp(used_at)

    @property
    def max(self):
        """The maximum energy."""
        return self._max

    @max.setter
    def max(self, max):
        """Configurates the maximum energy."""
        self.config(max=max)

    def _current(self, time=None):
        """Calculates the current internal energy.

        >>> energy = Energy(10, 300)
        >>> energy.use()
        >>> energy._current()
        9

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.
        """
        if not self.used:
            return self.max
        current = self.max - self.used + self.recovered(time)
        return current

    def current(self, time=None):
        """Calculates the current presentative energy. This equivalents to
        casting to ``int`` but can work with specified time.

        >>> energy = Energy(10, 300)
        >>> energy.use()
        >>> energy.current()
        9
        >>> int(energy)
        9

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.
        """
        return max(0, self._current(time))

    def debt(self, time=None):
        """Calculates the current energy debt.

        >>> energy = Energy(10, 300)
        >>> energy.debt()
        >>> energy.use(11, force=True)
        >>> energy.debt()
        1
        >>> energy.use(2, force=True)
        3

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.
        """
        current = self._current(time)
        if current >= 0:
            return
        return -current

    def use(self, quantity=1, time=None, force=False):
        """Consumes the energy.

        :param quantity: quantity of energy to be used. Defaults to ``1``.
        :param time: the time when using the energy. Defaults to the present
                     time in UTC.
        :param force: force to use energy even if there is not enough energy.
        :raise ValueError: not enough energy
        """
        time = timestamp(time)
        current = self._current(time)
        if current < quantity and not force:
            raise ValueError('Not enough energy')
        if current - quantity < self.max <= current or force:
            self.used = quantity - current + self.max
            self.used_at = time
        else:
            self.used = self.max - current + self.recovered(time) + quantity

    def recover_in(self, time=None):
        """Calculates seconds to the next energy recovery. If the energy is
        full or over the maximum, this returns ``None``.

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.
        """
        passed = self.passed(time)
        if passed is None or passed / self.recovery_interval >= self.used:
            return
        diff = self.recovery_interval - (passed % self.recovery_interval)
        current = self._current(time)
        if current < 0:
            return diff - current * self.recovery_interval
        return diff

    def recover_fully_in(self, time=None):
        """Calculates seconds to be recovered fully. If the energy is full or
        over the maximum, this returns ``None``.

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.

        .. versionadded:: 0.1.5
        """
        recover_in = self.recover_in(time)
        if recover_in is None:
            return
        to_recover = self.max - self.current()
        return recover_in + self.recovery_interval * (to_recover - 1)

    def recovered(self, time=None):
        """Calculates the recovered energy from the player used energy first.

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.
        """
        passed = self.passed(time)
        if passed is None:
            return 0
        recovered = (int(passed / self.recovery_interval) *
                     self.recovery_quantity)
        return min(recovered, self.used)

    def passed(self, time=None):
        """Calculates the seconds passed from using the energy first.

        :param time: the time when checking the energy. Defaults to the present
                     time in UTC.
        :raise ValueError: used at the future
        """
        if self.used_at is None:
            return
        seconds = timestamp(time) - self.used_at
        if seconds < 0:
            if self.future_tolerance is not None and \
               abs(seconds) <= self.future_tolerance:
                return 0
            raise ValueError('Used at the future (+%.2f sec)' % -seconds)
        return seconds

    def set(self, quantity, time=None):
        """Sets the energy to the fixed quantity.

        >>> energy = Energy(10, 300)
        >>> print energy
        <Energy 10/10>
        >>> energy.set(3)
        >>> print energy
        <Energy 3/10 recover in 05:00>

        You can also set over the maximum when give bonus energy.

        >>> energy.set(15)
        >>> print energy
        <Energy 15/10>

        :param quantity: quantity of energy to be set
        :param time: the time when setting the energy. Defaults to the present
                     time in UTC.
        """
        if quantity >= self.max:
            self.used = self.max - quantity
            self.used_at = None
        else:
            self.use(self.current(time) - quantity)

    def reset(self, time=None):
        """Makes the energy to be full. Most social games reset energy when the
        player reaches higher level.

        :param time: the time when setting the energy. Defaults to the present
                     time in UTC.
        """
        return self.set(self.max, time)

    def config(self, max=None, recovery_interval=None, time=None):
        """Updates :attr:`max` or :attr:`recovery_interval`.

        :param max: quantity of maximum energy to be set
        :param time: the time when setting the energy. Defaults to the present
                     time in UTC.
        """
        if max is not None:
            if self.recover_in(time):
                self.used += max - self._max
            self._max = max
        if recovery_interval is not None:
            self.recovery_interval = recovery_interval

    def __int__(self, time=None):
        """Type-casting to ``int``."""
        return self.current(time)

    def __float__(self, time=None):
        """Type-casting to ``float``."""
        return float(self.__int__(time))

    def __nonzero__(self, time=None):
        """Type-casting to ``bool``."""
        return bool(self.__int__(time))

    # Python 3 accepts __bool__ instead of __nonzero__
    __bool__ = __nonzero__

    def __eq__(self, other, time=None):
        """Is current energy equivalent to the operand.

        :param other: the operand
        :type other: :class:`Energy` or number
        """
        if isinstance(other, type(self)):
            return self.__getstate__() == other.__getstate__()
        elif isinstance(other, (int, float)):
            return float(self.current(time)) == other
        return False

    def __lt__(self, other, time=None):
        """Is current energy less than the operand.

        :param other: the operand
        :type other: number

        .. versionadded:: 0.1.3
        """
        return self.current(time) < other

    def __le__(self, other, time=None):
        """Is current energy less than or equivalent to the operand.

        :param other: the operand
        :type other: number

        .. versionadded:: 0.1.3
        """
        return self.current(time) <= other

    def __gt__(self, other, time=None):
        """Is current energy greater than the operand.

        :param number other: the operand
        :type other: number

        .. versionadded:: 0.1.3
        """
        return self.current(time) > other

    def __ge__(self, other, time=None):
        """Is current energy greater than or equivalent to the operand.

        :param other: the operand
        :type other: number

        .. versionadded:: 0.1.3
        """
        return self.current(time) >= other

    def __iadd__(self, other, time=None):
        """Increases by the operand.

        .. versionadded:: 0.1.1
        """
        self.set(self.current(time) + other, time)
        return self

    def __isub__(self, other, time=None):
        """Decreases by the operand.

        .. versionadded:: 0.1.1
        """
        return self.__iadd__(-other, time)

    def __getstate__(self):
        return {'used': self.used,
                'used_at': self.used_at,
                'max': self.max,
                'recovery_interval': self.recovery_interval,
                'recovery_quantity': self.recovery_quantity,
                'future_tolerance': self.future_tolerance}

    def __setstate__(self, state):
        if isinstance(state, tuple):
            # saved under 0.1.2
            self._max = state[0]
            self.recovery_interval = state[1]
            self.recovery_quantity = state[2]
            self.used = state[3]
            self.used_at = state[4]
            return
        self.used = state['used']
        self.used_at = state['used_at']
        self._max = state['max']
        self.recovery_interval = state['recovery_interval']
        self.recovery_quantity = state['recovery_quantity']
        self.future_tolerance = state['future_tolerance']

    def __repr__(self, time=None):
        current = self.current(time)
        rv = '<%s %d/%d' % (type(self).__name__, current, self.max)
        if current < self.max:
            recover_in = self.recover_in(time)
            rv += ' recover in %02d:%02d' % (recover_in / 60, recover_in % 60)
        return rv + '>'
